PwaIcon: Failed to store the icon image into the stream provided

### Description of the bug

When an exception occurs while [writing a PWA icon](https://github.com/vaadin/flow/blob/73a08b0805365b17218dccb5501d091d4b800834/flow-server/src/main/java/com/vaadin/flow/server/PwaIcon.java#L238) this exception is thrown uncontrolled until it's catched by a global error handler (or worst case: never).

This is happening quite often on our systems. The underlying error is mostly `` java.io.IOException: Broken pipe`` which happens e.g. when quickly reloading the site or aborting the site load.

<details><summary>Logs</summary>

```
2023-09-18 12:32:43 WARN  [http-nio-8080-exec-1] [omponents.error.ErrorUtil] Unhandled Exception detected!
2023-09-18 12:32:43 ERROR [http-nio-8080-exec-1] [omponents.error.ErrorUtil] Fatal Error
java.io.UncheckedIOException: Failed to store the icon image into the stream provided
        at com.vaadin.flow.server.PwaIcon.write(PwaIcon.java:240) ~[flow-server-24.1.11.jar!/:24.1.11]
        at com.vaadin.flow.server.communication.PwaHandler.lambda$init$1bc33a08$1(PwaHandler.java:83) ~[flow-server-24.1.11.jar!/:24.1.11]
        at com.vaadin.flow.server.communication.PwaHandler.handleRequest(PwaHandler.java:153) ~[flow-server-24.1.11.jar!/:24.1.11]
        at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1529) ~[flow-server-24.1.11.jar!/:24.1.11]
        at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:398) ~[flow-server-24.1.11.jar!/:24.1.11]
        at com.vaadin.flow.spring.SpringServlet.service(SpringServlet.java:106) ~[vaadin-spring-24.1.11.jar!/:?]
        at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:205) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationDispatcher.invoke(ApplicationDispatcher.java:642) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationDispatcher.processRequest(ApplicationDispatcher.java:408) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationDispatcher.doForward(ApplicationDispatcher.java:313) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationDispatcher.forward(ApplicationDispatcher.java:277) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.springframework.web.servlet.mvc.ServletForwardingController.handleRequestInternal(ServletForwardingController.java:141) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:178) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:51) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1081) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:974) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1011) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:903) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:564) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:205) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51) ~[tomcat-embed-websocket-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.springframework.security.web.ObservationFilterChainDecorator$FilterObservation$SimpleFilterObservation.lambda$wrap$1(ObservationFilterChainDecorator.java:479) ~[spring-security-web-6.1.3.jar!/:6.1.3]
        at org.springframework.security.web.ObservationFilterChainDecorator.lambda$wrapUnsecured$1(ObservationFilterChainDecorator.java:90) ~[spring-security-web-6.1.3.jar!/:6.1.3]
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:219) [spring-security-web-6.1.3.jar!/:6.1.3]
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:191) [spring-security-web-6.1.3.jar!/:6.1.3]
        at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:352) [spring-web-6.0.11.jar!/:6.0.11]
        at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:268) [spring-web-6.0.11.jar!/:6.0.11]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) [spring-web-6.0.11.jar!/:6.0.11]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) [spring-web-6.0.11.jar!/:6.0.11]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) [spring-web-6.0.11.jar!/:6.0.11]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) [spring-web-6.0.11.jar!/:6.0.11]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.springframework.web.filter.ServerHttpObservationFilter.doFilterInternal(ServerHttpObservationFilter.java:109) [spring-web-6.0.11.jar!/:6.0.11]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) [spring-web-6.0.11.jar!/:6.0.11]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) [spring-web-6.0.11.jar!/:6.0.11]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) [spring-web-6.0.11.jar!/:6.0.11]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:166) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:482) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:115) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:341) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:391) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:894) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1740) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-embed-core-10.1.12.jar!/:?]
        at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: org.apache.catalina.connector.ClientAbortException: java.io.IOException: Broken pipe
        at org.apache.catalina.connector.OutputBuffer.realWriteBytes(OutputBuffer.java:347) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.appendByteArray(OutputBuffer.java:750) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.append(OutputBuffer.java:677) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.writeBytes(OutputBuffer.java:382) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.write(OutputBuffer.java:360) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.CoyoteOutputStream.write(CoyoteOutputStream.java:97) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.CoyoteOutputStream.write(CoyoteOutputStream.java:90) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at com.vaadin.flow.server.PwaIcon.write(PwaIcon.java:238) ~[flow-server-24.1.11.jar!/:24.1.11]
        ... 67 more
Caused by: java.io.IOException: Broken pipe
        at sun.nio.ch.FileDispatcherImpl.write0(Native Method) ~[?:?]
        at sun.nio.ch.SocketDispatcher.write(Unknown Source) ~[?:?]
        at sun.nio.ch.IOUtil.writeFromNativeBuffer(Unknown Source) ~[?:?]
        at sun.nio.ch.IOUtil.write(Unknown Source) ~[?:?]
        at sun.nio.ch.IOUtil.write(Unknown Source) ~[?:?]
        at sun.nio.ch.SocketChannelImpl.write(Unknown Source) ~[?:?]
        at org.apache.tomcat.util.net.NioChannel.write(NioChannel.java:118) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper.doWrite(NioEndpoint.java:1381) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.net.SocketWrapperBase.doWrite(SocketWrapperBase.java:764) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.net.SocketWrapperBase.writeBlocking(SocketWrapperBase.java:589) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.net.SocketWrapperBase.write(SocketWrapperBase.java:533) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.coyote.http11.Http11OutputBuffer$SocketOutputBuffer.doWrite(Http11OutputBuffer.java:540) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.coyote.http11.filters.ChunkedOutputFilter.doWrite(ChunkedOutputFilter.java:112) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.coyote.http11.Http11OutputBuffer.doWrite(Http11OutputBuffer.java:193) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.coyote.Response.doWrite(Response.java:616) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.realWriteBytes(OutputBuffer.java:334) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.appendByteArray(OutputBuffer.java:750) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.append(OutputBuffer.java:677) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.writeBytes(OutputBuffer.java:382) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.write(OutputBuffer.java:360) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.CoyoteOutputStream.write(CoyoteOutputStream.java:97) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.CoyoteOutputStream.write(CoyoteOutputStream.java:90) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at com.vaadin.flow.server.PwaIcon.write(PwaIcon.java:238) ~[flow-server-24.1.11.jar!/:24.1.11]
        ... 67 more
```

</details>

### Expected behavior

The "error" can neither be fixed nor handled and causes a lot of log spamming, so it would be nice if you could catch it beforehand and log it to e.g. debug level.

### Minimal reproducible example

This exception is hard to reproduce, as it only happens e.g. when the client cuts the connection, which is not easily controllable.

To reproduce the problem I recommend downloading the Vaadin starter, adding PWA support and hitting the refresh button very often.

### Versions

- Vaadin / Flow version: 24.1.9 (currently latest version)

--- a/flow-server/src/main/java/com/vaadin/flow/server/communication/PwaHandler.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/communication/PwaHandler.java
@@ -86,7 +86,7 @@ public class PwaHandler implements RequestHandler {
                             icon.write(out);
                         } catch (UncheckedIOException ex) {
                             LoggerFactory.getLogger(PwaHandler.class)
-                                    .debug("Error serving PWA image", ex);
+                                    .debug("Error serving PWA icon", ex);
                         }
                         return true;
                     });

--- a/flow-server/src/main/java/com/vaadin/flow/server/communication/PwaHandler.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/communication/PwaHandler.java
@@ -18,11 +18,14 @@ package com.vaadin.flow.server.communication;
 import java.io.IOException;
 import java.io.OutputStream;
 import java.io.PrintWriter;
+import java.io.UncheckedIOException;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+
+import org.slf4j.LoggerFactory;
 
 import com.vaadin.flow.function.SerializableSupplier;
 import com.vaadin.flow.server.PwaIcon;
@@ -81,6 +84,9 @@ public class PwaHandler implements RequestHandler {
                         }
                         try (OutputStream out = response.getOutputStream()) {
                             icon.write(out);
+                        } catch (UncheckedIOException ex) {
+                            LoggerFactory.getLogger(PwaHandler.class)
+                                    .debug("Error serving PWA image", ex);
                         }
                         return true;
                     });

--- a/flow-server/src/test/java/com/vaadin/flow/server/communication/PwaHandlerTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/communication/PwaHandlerTest.java
@@ -36,11 +36,11 @@ import com.vaadin.flow.server.VaadinSession;
 
 public class PwaHandlerTest {
 
-    private VaadinSession session = Mockito.mock(VaadinSession.class);
+    private final VaadinSession session = Mockito.mock(VaadinSession.class);
 
-    private VaadinRequest request = Mockito.mock(VaadinRequest.class);
+    private final VaadinRequest request = Mockito.mock(VaadinRequest.class);
 
-    private VaadinResponse response = Mockito.mock(VaadinResponse.class);
+    private final VaadinResponse response = Mockito.mock(VaadinResponse.class);
 
     @Test
     public void handleRequest_noPwaRegistry_returnsFalse() throws IOException {

--- a/flow-server/src/test/java/com/vaadin/flow/server/communication/PwaHandlerTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/communication/PwaHandlerTest.java
@@ -18,12 +18,17 @@ package com.vaadin.flow.server.communication;
 import java.io.IOException;
 import java.io.PrintWriter;
 import java.io.StringWriter;
+import java.io.UncheckedIOException;
+import java.lang.reflect.Constructor;
+import java.util.List;
 
 import org.junit.Assert;
 import org.junit.Test;
+import org.mockito.ArgumentMatchers;
 import org.mockito.Mockito;
 
 import com.vaadin.flow.server.PwaConfiguration;
+import com.vaadin.flow.server.PwaIcon;
 import com.vaadin.flow.server.PwaRegistry;
 import com.vaadin.flow.server.VaadinRequest;
 import com.vaadin.flow.server.VaadinResponse;
@@ -92,5 +97,41 @@ public class PwaHandlerTest {
         Assert.assertTrue(handler.handleRequest(session, request, response));
 
         Mockito.verify(registry, Mockito.times(1)).getIcons();
+    }
+
+    @Test
+    public void handleRequest_writeIconOnResponseFailure_doesNotThrow()
+            throws Exception {
+
+        PwaRegistry registry = Mockito.mock(PwaRegistry.class);
+        PwaConfiguration configuration = Mockito.mock(PwaConfiguration.class);
+        Mockito.when(registry.getPwaConfiguration()).thenReturn(configuration);
+
+        PwaHandler handler = new PwaHandler(() -> registry);
+
+        PwaIcon icon = Mockito.spy(createIcon(registry, 100));
+        Mockito.when(registry.getIcons()).thenReturn(List.of(icon));
+        Mockito.when(configuration.isEnabled()).thenReturn(true);
+
+        Mockito.doAnswer(i -> {
+            throw new UncheckedIOException(
+                    "Failed to store the icon image into the stream provided",
+                    new IOException("Broken pipe"));
+        }).when(icon).write(ArgumentMatchers.any());
+        Mockito.when(request.getPathInfo())
+                .thenReturn("/icons/icon-100x100.png");
+
+        Assert.assertTrue(handler.handleRequest(session, request, response));
+    }
+
+    private PwaIcon createIcon(PwaRegistry registry, int size)
+            throws Exception {
+        Constructor<PwaIcon> ctor = PwaIcon.class
+                .getDeclaredConstructor(int.class, int.class, String.class);
+        ctor.setAccessible(true);
+        PwaIcon icon = ctor.newInstance(size, size,
+                PwaConfiguration.DEFAULT_ICON);
+        icon.setRegistry(registry);
+        return icon;
     }
 }

